### ISSUE TITLE
twisted.python.filepath.FilePath.open sometimes returns a file in text mode, sometimes in binary mode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4453]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4453|
|Type|defect|
|Created|2010-05-24 04:01:06Z|
|Branch|https://github.com/twisted/twisted/tree/o_binary-4453|

As discovered in [#3694](https://github.com/twisted/twisted/issues/3694), the text/binary mode of the file returned by `FilePath.open` varies based on whether `requireCreate` has been called on the `FilePath` or not.

This makes it difficult to actually write the data you're expecting to write, since on Windows the text/binary mode of the file will actually change the bytes that end up on disk.

Also, the mode of the file object returned from `path.open()` after `path.requireCreate()` says "w+b".  The "b" is a lie.  It's open in text mode.

Attachments:

* [openbug.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/812/8121b27f2e984dee362dbbcc12360a43d4756453/af1b359a32d1b8e962d87b7b4e5c12d054ba2535.py) (314 bytes) - added by exarkun on 2010-05-24 04:01:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4453 4453
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_o_binary_4453 branches/o_binary-4453
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1274673666000000 1274673666000000
changetime__1274687114000000 1274687114000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [29023]) Branching to 'o_binary-4453'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

[Build results.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/o_binary-4453)

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi set owner to @glyph|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set status to `closed`|
|-|-|

(In [29027]) Merge o_binary-4453: always open files in binary mode, even when using create()

Author: glyph

Reviewer: mithrandi

Fixes: [#4453](https://github.com/twisted/twisted/issues/4453)

Files opened with FilePath.create() (and therefore also files opened via FilePath.open() on a path with alwaysCreate=True) will now be opened in binary mode as advertised, so that they will behave portably across platforms.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
